torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -0,0 +1,406 @@
+    const ::gloo::Exception& ex) {
+  {
+    std::unique_lock<std::mutex> lock(m_);
+    completed_ = true;
+    ex_ = std::unique_ptr<::gloo::Exception>(new ::gloo::Exception(ex));
+  }
+  cv_.notify_all();
+}
+
+ProcessGroupGloo::Options::Options()
+    : timeout(std::chrono::milliseconds(10 * 1000)), threads(1) {}
+
+ProcessGroupGloo::ProcessGroupGloo(
+    const std::shared_ptr<Store>& store,
+    int rank,
+    int size)
+    : ProcessGroup(rank, size), store_(new GlooStore(store)), stop_(false) {}
+
+ProcessGroupGloo::~ProcessGroupGloo() {
+  // Require this process group to be explicitly shut down prior to being
+  // destructed. This is the easiest way to guarantee clean shutdown
+  // and avoid blocking/throwing in a destructor.
+}
+
+void ProcessGroupGloo::initialize() {
+  Options options;
+  options.timeout = std::chrono::milliseconds(100);
+  options.threads = 2;
+  initialize(options);
+}
+
+void ProcessGroupGloo::initialize(Options& options) {
+  auto& devices = options.devices;
+  if (devices.empty()) {
+    devices.push_back(::gloo::transport::tcp::CreateDevice("localhost"));
+  }
+
+  for (auto& device : options.devices) {
+    auto context = std::make_shared<::gloo::rendezvous::Context>(rank_, size_);
+    context->setTimeout(options.timeout);
+    context->connectFullMesh(*store_, device);
+    contexts_.push_back(std::move(context));
+  }
+
+  threads_.resize(options.threads);
+  for (int i = 0; i < threads_.size(); i++) {
+    threads_[i] = std::thread(&ProcessGroupGloo::runLoop, this);
+  }
+}
+
+void ProcessGroupGloo::destroy() {
+  std::unique_lock<std::mutex> lock(m_);
+  while (!queue_.empty()) {
+    queueConsumeCV_.wait(lock);
+  }
+
+  // Queue is empty, signal stop
+  stop_ = true;
+
+  // Release lock to allow threads to terminate
+  queueProduceCV_.notify_all();
+  lock.unlock();
+
+  // Wait for worker threads to terminate
+  for (auto& thread : threads_) {
+    thread.join();
+  }
+}
+
+void ProcessGroupGloo::runLoop(void) {
+  std::unique_lock<std::mutex> lock(m_);
+
+  while (!stop_) {
+    if (queue_.empty()) {
+      queueProduceCV_.wait(lock);
+      continue;
+    }
+
+    auto tuple = std::move(queue_.front());
+    queue_.pop_front();
+    queueConsumeCV_.notify_one();
+
+    auto& entry = std::get<0>(tuple);
+    auto& work = std::get<1>(tuple);
+
+    // Continue holding onto the lock; this ensures that we serialize
+    // creation of Gloo algorithm instances for the context associated
+    // with this process group.
+    if (!entry->algorithm) {
+      createAlgorithm(*entry);
+    }
+
+    lock.unlock();
+    runSingle(std::move(tuple));
+    lock.lock();
+  }
+}
+
+void ProcessGroupGloo::runSingle(WorkType tuple) {
+  auto& entry = std::get<0>(tuple);
+  auto& work = std::get<1>(tuple);
+
+  try {
+    entry->run(entry);
+    work->finish();
+  } catch (const ::gloo::Exception& ex) {
+    work->finishWithException(ex);
+  }
+
+  // Return entry to cache
+  std::unique_lock<std::mutex> lock(m_);
```

Comment:
Could we pass in the reference to `unique_lock` into this function, and avoid unlocking the mutex just to acquire it after returning?